### PR TITLE
[WIP] Imperative Programming Scaffold

### PR DESCRIPTION
**Pull request by cs2be**
Please refer to https://github.com/PaddlePaddle/Paddle/issues/9912.  Add in initial design doc for Imperative Programming.  Add scaffold for the imperative Reader.

### CODE REVIEW

File: doc/fluid/design/imperative/imperative.md [hunk 1]
```diff
@@ -0,0 +1,99 @@
...
+may create block id and variable names that may conflict with each other
+when we merge the programs together.  We must take special care when 
+re-indexing the block ids and merging variables.
+
+### Training Loop
+
+With imperative, the training loop will be added to the main program as an 
+operator.  Since a majority of the use cases involves reading in training data
+and feeding the data to the network to be trained, we chose to create an 
+iterate method on reader to act as the training loop.
```
Comment (tonyyang-svail) [inline doc/fluid/design/imperative/imperative.md]: What is the advantage of writing a specialized `ReaderOp` over writing a simple `IteratorOp`
```python
with Iterator(max_iter=1000):
    x, y = reader()
    ...
```

File: doc/fluid/design/imperative/imperative.md [hunk 2]
```diff
@@ -0,0 +1,99 @@
...
+
+This iterate method will loop through the data until it has reached EOF.  
+During each iteration, it will fetch training data and put them in user defined
+variables.  The block within the with reader.iterate() method represents the
+network to be trained.
+
+We will need to create a `reader_loop_op`, which implement the training loop.
+
+```
+READER_FILE_PATH =  './data.recordio'
```
Comment (helinwang) [inline doc/fluid/design/imperative/imperative.md]: Maybe add Python at the end for a Python syntax highlight ("```Python")

File: doc/fluid/design/imperative/imperative.md [hunk 3]
```diff
@@ -0,0 +1,99 @@
...
+READER_BATCH_SIZE = 128
+TRAIN_LOOP_STEPS = 100
+MODEL_DIR = './model'
+
+reader = fluid.batch_reader(file=READER_FILE_PATH,
+                                               batch_size=READER_BATCH_SIZE,
```
Comment (helinwang) [inline doc/fluid/design/imperative/imperative.md]: Maybe fewer spaces for readability.

File: doc/fluid/design/imperative/imperative.md [hunk 4]
```diff
@@ -0,0 +1,99 @@
...
+  sgd_optimizer.minimize(avg_cost)
+
+fluid.save_model(dir=MODEL_DIR)
+``` 
+
+### File Readers
+
+We can utilize our existing recordio reader, batch reader, and shuffle reader 
+ops, however we will need to make some modifications to support methods to 
+check for EOF (and any additional methods used by our `reader_loop_op`)
```
Comment (helinwang) [inline doc/fluid/design/imperative/imperative.md]: I think the current reader interface already provides a way to check for EOF (the `has_next` method).

File: doc/fluid/design/imperative/imperative.md (hunk 1 above)
Comment (cs2be) [reply doc/fluid/design/imperative/imperative.md]: Iterator op would introduce a general purpose control flow operator like While or IFElse, which are control flow operators that can be used for training purposes since they support back prop.

The ReaderOp iterator is specifically used for training loop (which should not be used for back prop).  If we introduce a general purpose op to iterate, we feel it would be confusing for users, since it does not behave like the other control flow ops.

File: doc/fluid/design/imperative/imperative.md [hunk 5]
```diff
@@ -0,0 +1,111 @@
...
+TRAIN_LOOP_STEPS = 100
+MODEL_DIR = './model'
+
+reader = fluid.batch_reader(file=READER_FILE_PATH, 
+                            batch_size=READER_BATCH_SIZE,
+                            shape=[[13], [1]], 
+                            dtype=['float32', 'float32'],
+                            format='recordio')
+
+with reader.iterate() as (x, y):
```
[the remark this answers is not in the document]
Comment (cs2be) [reply doc/fluid/design/imperative/imperative.md]: This method will actually add an operator to program desc and will be looped by the exectuor

File: doc/fluid/design/imperative/imperative.md [hunk 6]
```diff
@@ -0,0 +1,111 @@
...
+and Paddle Fluid API.  To further complicate the code, the neural network logic
+could be combined with the run logic (ie: Logic to run on a distributed 
+system).
+
+The mix of Python and Paddle API also makes it impossible for a transpiler to
+convert the ProgramDesc to C++.
+
+## Changes to support Imperative
+
+Please refer to [Paddle API v4 Proposal](https://github.com/PaddlePaddle/Paddle/issues/9912)
```
[the remark this answers is not in the document]
Comment (cs2be) [reply doc/fluid/design/imperative/imperative.md]: This is not true imperative because we do not have eager execution.  This change is about moving the all Computation Representation to the IR (ProgramDesc).  Currently users mix in Python code with IR (ie: load training data in python, run executor manually for startup and main program), which makes it difficult to transpile.

File: doc/fluid/design/imperative/imperative.md [hunk 7]
```diff
@@ -0,0 +1,111 @@
...
+methods used by our `reader_loop_op`)
+
+### Save Model
+
+Current a user is able to save trained parameters by calling
+`fluid.io.save_persistables`.  This method will create a new Program, which
+will save all the parameters to a file.  With imperative fluid, we no longer 
+need to create a new Program to save variables, however we can create a simple
+helper method `fluid.save_model` that will add `save`/`save_combined` operators
+in the main program.
```
[the remark this answers is not in the document]
Comment (cs2be) [reply doc/fluid/design/imperative/imperative.md]: save_model will call our current save_persistables.  The training loop is within "with reader.iterate() as (x, y):", we save the model after the loop is complete.

File: doc/fluid/design/imperative/imperative.md [hunk 8]
```diff
@@ -0,0 +1,111 @@
+# Imperative Programming
```
[the remark this answers is not in the document]
Comment (cs2be) [reply doc/fluid/design/imperative/imperative.md]: This is a misnomer, perhaps we can rename this to "Simplify Paddle API to support Transpilation".  As explained above, we won't be able to support imperative unless we go the pytorch route.  I'll discuss with team to get this renamed.

File: doc/fluid/design/imperative/imperative.md (hunk 1 above)
Comment (tonyyang-svail) [reply doc/fluid/design/imperative/imperative.md]: Sounds good. Thanks for the explanation.
